File: 2021/w38/kata_go_bank-kata/core/model/account/account-service.go
package account

import (
	"time"

	"kata/core/model"
)

type AccountService struct {
	repo model.AccountRepo

	TimeSource func() time.Time
}

func NewAccountService(repo model.AccountRepo) *AccountService {
	res := &AccountService{
		repo:       repo,
		TimeSource: func() time.Time { return time.Now().UTC() },
	}

	return res
}

func (obj *AccountService) Deposit(clientID model.ClientID, amount model.Amount) error {
	return obj.repo.Save(model.TransactionRecord{
		ClientID:        clientID,
		TransactionType: model.Deposit,
		Amount:          amount,
		Date:            obj.TimeSource(),
	})
}

func (obj *AccountService) Withdrawal(clientID model.ClientID, amount model.Amount) error {
	return obj.repo.Save(model.TransactionRecord{
		ClientID:        clientID,
		TransactionType: model.Withdrawal,
		Amount:          amount,
		Date:            obj.TimeSource()})
}

func (obj *AccountService) PrintStatement(clientID model.ClientID) (*model.Statement, error) {
	records, err := obj.repo.Load(clientID)
	if err != nil {
		return nil, err
	}

	return model.NewStatement(records), nil
}

File: 2022/w04/bank_kata_ex/lib/bank/core/model/account-service.ex
defmodule Bank.Core.Model.Account.Service do
  @moduledoc false

  @behaviour Bank.Core.Model.Account.Boundaries.Service

  alias Bank.Core.Model.Amount
  alias Bank.Core.Model.Account

  @spec deposit(account_id: String.t(), amount: Amount.t(), time: NaiveDateTime.t()) ::
          :ok | {:error, any()}
  def deposit(account_id: account_id, amount: %Amount{} = amount, time: %NaiveDateTime{} = time) do
    with {:ok, account} <- account_repo().load(account_id),
         {:ok, account} <- Account.deposit(account: account, amount: amount, time: time) do
      account_repo().save(account)
    else
      error ->
        {:error, error}
    end
  end

  @spec withdraw(account_id: String.t(), amount: Amount.t(), time: NaiveDateTime.t()) ::
          :ok | {:error, any()}
  def withdraw(account_id: account_id, amount: %Amount{} = amount, time: %NaiveDateTime{} = time) do
    with {:ok, account} <- account_repo().load(account_id),
         {:ok, account} <- Account.withdraw(account: account, amount: amount, time: time) do
      account_repo().save(account)
    else
      error ->
        {:error, error}
    end
  end

  @spec print_statement(account_id: String.t()) :: {:ok, String.t()} | {:error, any()}
  def print_statement(account_id: account_id) do
    with {:ok, account} <- account_repo().load(account_id),
         {:ok, statement} <- Account.print_statement(account: account) do
      {:ok, statement}
    else
      error ->
        {:error, error}
    end
  end

  defp account_repo do
    Application.get_env(:bank_kata_ex, :repos) |> Map.get(:account)
  end
end
